### PR TITLE
Add `no-hooks-for-single-case` rule (fixes #44)

### LINKED ISSUE #44: Rule proposal: disallow before, beforeEach, after, afterEach if you have only one test

### PR DESCRIPTION
**Pull request by jfmengels**
Add `no-hooks-for-single-case` rule (fixes #44)

### CODE REVIEW

**Comment by lo1tuma** [comment]
It turns out that there is one good use-case for using the `after` hook even if there is only one test (which I wasn’t aware of when I created #44).

If a test requires some clean-up code to run even if the test fails then this must be done in the `after` hook because the test function exits on the first failing assertion.

File: lib/rules/no-hooks-for-single-case.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    function popLayer(node) {
+        var layer = layers[layers.length - 1];
+        if (layer.describeNode === node) {
+            if (layer.testCount <= 1) {
+                layer.hookNodes.forEach(function (hookNode) {
+                    context.report({
+                        node: hookNode,
+                        // message: 'Unexpected use of duplicate Mocha `' + name + '` hook'
+                        message: 'Unexpected use of duplicate Mocha `before` hook'
```
Comment (lo1tuma) [inline lib/rules/no-hooks-for-single-case.js]: That message doesn’t seem to make sense for this rule. Shouldn’t be something like "Unnecessary use of Mocha hook for only one test case"?

Comment (jfmengels) [inline lib/rules/no-hooks-for-single-case.js]: Absolutely. I simply forgot to add a new message... :sweat_smile: (I'll remove the comment too)

**Comment by jfmengels** [comment]
> If a test requires some clean-up code to run even if the test fails then this must be done in the after hook because the test function exits on the first failing assertion.

True.
Do you think `after` should always be allowed, behind a flag (enabled or not by default?). Same thing for `afterEach`?

**Comment by lo1tuma** [comment]
I would simply always allow to use `after`; `afterEach` could be simply renamed to `after` for a single test case. But maybe should rethink the name of rule. What do you think about `no-unnecessary-hooks`?

**Comment by jfmengels** [comment]
`no-unnecessary-hooks` doesn't fit the bill IMO. If you have a `before` for one single case, the hook is not "unnecessary", it's just that we want to promote an other way of doing it.
`no-setup-for-single-case` / `no-before-for-single-case` would work, but wouldn't catch the `afterEach` meaning. In the direction of `no-common-code-for-single-case` maybe?

I'm fine with the current name, but I don't mind bikeshedding some more if you think it doesn't fit.

**Comment by lo1tuma** [comment]
For `no-hooks-for-single-case` it would be strange to make exceptions by default, that’s why I thought it might be good to find a rule name that doesn’t give the impression that all hooks are forbidden for a single test case. Maybe we can simply replace `no` with `avoid`, which would be less strict?

**Comment by jfmengels** [comment]
Hey, sorry for not replying, I was on vacation :)

> For no-hooks-for-single-case it would be strange to make exceptions by default

I don't think it is that strange, but I agree it could be confusing (but that's why there is documentation). We could forbid `after` by default but push for users to configure it by allowing `after`. I think that is probably the least confusing option. Maybe then have a new `allow` option which takes an array of strings: `['after', 'afterEach']`?

I'd stick with `no-` as that is the convention with ESLint rules, even with similar cases where not everything gets forbidden using options.

**Comment by lo1tuma** [comment]
@jfmengels No problem, I hope you had a great time ;).

Ok, so let’s go with your suggestion and forbid all hooks per default but allow to define exceptions per option.

**Comment by jfmengels** [comment]
> I hope you had a great time ;)

I did :)

> Ok, so let’s go with your suggestion and forbid all hooks per default but allow to define exceptions per option.

:+1:, I'll try to get on this soon. Will fix the other things you mentioned in the same go, like the error message.

**Comment by jfmengels** [comment]
Added an `allow` option, and updated the error message.

File: docs/rules/no-hooks-for-single-case.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    describe('bar', function() {
+      /* ... */
+    });
+});
+```
+
+# Options
+
+This rule supports the following shared configuration options:
```
Comment (lo1tuma) [inline docs/rules/no-hooks-for-single-case.md]: `allow` is a rule option not a shared configuration, right?

File: docs/rules/no-hooks-for-single-case.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+* `allow`: An array containing the names of hooks to allow. This might be used to allow writing `after` hooks to run clean-up code. Defaults to an empty array.
+
+```json
+{
+    "rules": {
+        "mocha/no-skipped-tests": ["error", {"allow": ["after"]}]
```
Comment (lo1tuma) [inline docs/rules/no-hooks-for-single-case.md]: Wrong rule name.

**Comment by jfmengels** [comment]
Oops... good catch. Fixed.

**Comment by lo1tuma** [comment]
Awesome. Thank you!

**Comment by jfmengels** [comment]
My pleasure. That was the last one of the rules that I wrote :smile:
I might take a look at working on others later, there are still some interesting ones in discussion.

**Pull request merged**